### PR TITLE
Add StartUnit operation

### LINKED ISSUE #21: Move from POC: [Node] function to start a unit
## Please describe what you would like to see

The dbus connection from https://github.com/engelmi/orch/issues/20 is used to start units. 

## Please describe the solution you'd like

For reference, see [the implementation from the POC](https://github.com/alexlarsson/orch/blob/main/node.c#L166-L190). 
The callback for the systemd start unit call can be empty or just print out a simple message (for now).
Maybe adding a simple unit test (just a dedicated `<xyz_test.c` file containing a `main()`) can help to verify that the systemd unit can be started. 

_Important Tech note:_ Instead of `"isolate"` use `"replace"`. This is the default behavior of `systemctl start` (see [here](https://www.freedesktop.org/software/systemd/man/systemctl.html#--job-mode=) for reference) and should be used instead of `isolate` (as it causes all other units to be stopped). 

image

### PR DESCRIPTION
**Pull request by alexlarsson**
This calls StartUnit on the agent, and if that succeeds it signals
the availability of a new job object and returns its object path.
    
Eventually the agent will detect the matching JobRemoved signal from
systemd and will send a JobDone to the manager, which will then
emit a JobRemoved on the hirte job object.
    
Compared to the dbus API document some things are missing:
 * There is no queueing of conflicting operations
 * We don't support properties on the Job object.

### CODE REVIEW

**Comment by alexlarsson** [comment]
This is initial work on https://github.com/containers/hirte/issues/91

File: src/agent/agent.c [hunk 1]
```diff
@@ -12,6 +12,21 @@
...
 
+struct JobTracker {
+        char *job_object_path;
+        job_tracker_callback callback;
+        void *userdata;
+        free_func_t free_userdata;
+        LIST_FIELDS(JobTracker, tracked_jobs);
+};
+
+static bool
```
Comment (sdunnagan) [inline src/agent/agent.c:23]: Extra white space here.

File: src/agent/agent.c [hunk 2]
```diff
@@ -263,13 +289,175 @@ static int agent_method_list_units(UNUSED sd_bus_message *m, void *userdata, UNU
...
+        if (trackp && *trackp) {
+                job_tracker_free(*trackp);
+                *trackp = NULL;
+        }
+}
+
+#define _cleanup_job_tracker_ _cleanup_(job_tracker_freep)
+
+static bool
+                agent_track_job(Agent *agent,
```
Comment (sdunnagan) [inline src/agent/agent.c:406]: Extra white space here.

File: src/agent/agent.c [hunk 3]
```diff
@@ -263,13 +289,175 @@ static int agent_method_list_units(UNUSED sd_bus_message *m, void *userdata, UNU
...
+        const char *unit = NULL;
+        const char *result = NULL;
+        JobTracker *track = NULL, *next_track = NULL;
+        uint32_t id = 0;
+        int r = 0;
+
+        r = sd_bus_message_read(m, "uoss", &id, &job_path, &unit, &result);
+        if (r < 0) {
+                fprintf(stderr, "Can't parse job result\n");
+                return 0;
```
Comment (sdunnagan) [inline src/agent/agent.c]: ```r``` should be returned here rather than 0.

File: src/manager/job.c [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        static uint32_t next_id = 0;
+
+        _cleanup_job_ Job *job = malloc0(sizeof(Job));
+        if (job == NULL) {
+                return NULL;
+        }
+
+        job->ref_count = 1;
+        job->state = JOB_WAITING;
+        job->id = ++next_id;
```
Comment (sdunnagan) [inline src/manager/job.c:18]: This can just be set to 1.

**Review by sdunnagan — APPROVED** [review]

File: src/agent/agent.c (hunk 1 above)
Comment (alexlarsson) [reply src/agent/agent.c:23]: This is added by out clang formater, so not sure what I can do about it.

File: src/agent/agent.c (hunk 2 above)
Comment (alexlarsson) [reply src/agent/agent.c:406]: same here

File: src/manager/job.c (hunk 4 above)
Comment (alexlarsson) [reply src/manager/job.c:18]: What do you mean one? We need a unique id for each job.

File: src/agent/agent.c (hunk 3 above)
Comment (alexlarsson) [reply src/agent/agent.c]: yeah

File: src/manager/job.c (hunk 4 above)
Comment (sdunnagan) [reply src/manager/job.c:18]: Ah, I missed ```next_id``` is static.

**Pull request merged**